internal/pkg/crypto/ca.go

```diff
@@ -0,0 +1,133 @@
+	}
+	err = pem.Encode(publicKeyPEM, &pem.Block{
+		Type:  "PUBLIC KEY",
+		Bytes: publicKey,
+	})
+	if err != nil {
+		return "", errors.Join(errors.New("cannot encode public key"), err)
+	}
+	return publicKeyPEM.String(), nil
+}
+
+func GenerateCA() (*CA, error) {
+	privateKey, err := newPrivateKey(1024)
+	if err != nil {
+		return nil, errors.Join(errors.New("cannot create private key"), err)
+	}
+	serialNumber, err := rand.Int(rand.Reader, (&big.Int{}).Exp(big.NewInt(2), big.NewInt(159), nil))
+	if err != nil {
+		return nil, errors.Join(errors.New("cannot init serial number"), err)
+	}
+	caCertificateTemplate := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Organization:  []string{""},
+			Country:       []string{""},
+			Province:      []string{""},
+			Locality:      []string{""},
+			StreetAddress: []string{""},
+			PostalCode:    []string{""},
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0), // 10 years expiration
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	caCertificateBytes, err := x509.CreateCertificate(
+		rand.Reader,
+		&caCertificateTemplate,
+		&caCertificateTemplate,
+		&privateKey.Key().PublicKey,
+		privateKey.Key())
+	if err != nil {
+		return nil, errors.Join(errors.New("cannot create certificate"), err)
+	}
+	return &CA{caCertificateTemplate, caCertificateBytes, privateKey.Key()}, nil
+}
+
+// ExtractCaFromSecret generates a CA object from the given secret where the
+// certificate and private key are stored
+func ExtractCaFromSecret(caSecret *corev1.Secret) (*CA, error) {
+	caCertBytes, hasCARootCert := caSecret.Data[corev1.TLSCertKey]
+	if !hasCARootCert {
+		return nil, fmt.Errorf("CA could not be extracted from secret %s", caSecret.Kind)
+	}
+	caCertificateBytes, _ := pem.Decode(caCertBytes)
+	if caCertificateBytes == nil {
+		return nil, fmt.Errorf("failed to decode Root CA certificate")
+	}
+	caCert, err := x509.ParseCertificate(caCertificateBytes.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("CA certificate could not be extracted from secret %s", caSecret.Kind)
+	}
+
+	caPrivateKeyBytes, hasCARootCert := caSecret.Data[corev1.TLSPrivateKeyKey]
+	if !hasCARootCert {
+		return nil, fmt.Errorf("CA private key bytes could not be extracted from secret %s", caSecret.Kind)
+	}
+	block, _ := pem.Decode(caPrivateKeyBytes)
+	if block == nil {
+		return nil, fmt.Errorf("failed to decode root ca private key")
+	}
+
+	caPrivateKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("CA private key could not be extracted from secret %s", caSecret.Kind)
+	}
+
+	return &CA{CaCert: *caCert, CaCertBytes: caCertificateBytes.Bytes, CaPrivateKey: caPrivateKey}, nil
+}
```

Comment:
I'm aware that this function is very similar to another in the `internal/pkg/admission/policy-server-ca-secret.go`. I have not remove that function because that require further changes in the reconciliation loop to keep  the tests passing and I would like to keep this PR as small as possible. This reconciliation loop changes will be the next step.